##  Conjugate Gradient Method_2:

\\begin{equation}
min \ f(x),\ f(x) = 0.5x^T A x - b^T x
\\end{equation}

\begin{eqnarray}
\ A = \begin{pmatrix}
1 & 0 & 0 \\
0 & 5 & 0 \\
0 & 0 & 25 \\
\end{pmatrix}
\end{eqnarray}

\\begin{equation}
\ Using \ b^T = [-1,-1,-1]
\\end{equation}

In [90]:
import math
import numpy as np

**Initializing the vectors**

In [91]:
x0 = np.matrix([[0],[0],[0]])
A  = np.matrix([[1,0,0],[0,5,0],[0,0,25]]) 
b =  np.matrix([[-1],[-1],[-1]])

**The Function to compute the minimizer vector by Conjugate gradient method**

In [109]:
def conjugate_gradient(A, b, x0, tol = 1.0e-8, max_iter = 100):
    """
    A function to solve [A]{x} = {b} linear equation system with the 
    conjugate gradient method.
    
    :param A : array 
        A real symmetric positive definite matrix(assumed)
        
    :param b : vector
        The vector of the system which is given in RHS.
        
    :param x0 : vector
        The starting guess for the solution.
        
    :param max_iter : integer
        Maximum number of iterations. Iteration will stop after max_iter 
        steps even if the specified tolerance has not been achieved.
        
    :param tol : float
        Tolerance to achieve. The algorithm will terminate when either 
        the relative or the absolute residual is below tol.
        
    :var    r0 : vector
                 Initialization stores the value (b - a * A )
    
    :var    d  : vector
    
    :var    a  : float
                 Iteratively computes the scalar of (r1T.r1)/(r0T.r0)

    :var    ri : vector
                 Iteratively stores the value (r - a * A * d), used to check for the convergence
    
    :var    x  : vector 
                 Stores the solution for the next iteration iteratively
                 
    :var    b  : float
                 Iteratively computes the scalar of (riT.ri)/(diT.A.di)
    """
    x = x0
    r0 = b - np.dot(A, x)
    d = r0
    
    def function_value(x):
        return round(np.dot(0.5 * x.T,np.dot(A,x)).item(0) - np.dot(b.T,x).item(0),4)
    
#   Iterations:   
    for i in xrange(max_iter):
        a = float(np.dot(r0.T, r0)/np.dot(np.dot(d.T, A), d))
        x = x + d*a
        ri = r0 - np.dot(A*a, d)
        
        print "iteration: ",i, "r(i): ",round(np.linalg.norm(ri),5),"xi :(",round(x.item(0),4),",",round(x.item(1),4),",",round(x.item(2),4),")","F(xi): ",function_value(x)
        
        if np.linalg.norm(ri) < tol:
            print "\nConverged Successfully in iterations :",i
            print "The result of vector x:"
            return np.around(x,decimals=10)
            break
        b2 = float(np.dot(ri.T, ri)/np.dot(r0.T, r0))
        d = ri + b2 * d
        r0 = ri
    return x

In [110]:
conjugate_gradient(A, b, x0, tol = 1.0e-08, max_iter = 100)

iteration:  0 r(i):  1.73205 xi :( 0.0 , 0.0 , 0.0 ) F(xi):  0.0
iteration:  1 r(i):  1.23458 xi :( -0.0484 , -0.0484 , -0.0484 ) F(xi):  -0.1089
iteration:  2 r(i):  1.42683 xi :( -0.132 , -0.1238 , -0.0827 ) F(xi):  -0.206
iteration:  3 r(i):  1.3112 xi :( -0.2742 , -0.2356 , -0.0831 ) F(xi):  -0.3302
iteration:  4 r(i):  0.86392 xi :( -0.3714 , -0.2923 , -0.0549 ) F(xi):  -0.3983
iteration:  5 r(i):  0.82692 xi :( -0.4421 , -0.3072 , -0.0283 ) F(xi):  -0.4339
iteration:  6 r(i):  1.14915 xi :( -0.5851 , -0.3014 , -0.0022 ) F(xi):  -0.4904
iteration:  7 r(i):  1.00087 xi :( -0.8089 , -0.2721 , -0.0034 ) F(xi):  -0.5721
iteration:  8 r(i):  0.47564 xi :( -0.9059 , -0.2518 , -0.0245 ) F(xi):  -0.6059
iteration:  9 r(i):  0.23975 xi :( -0.9275 , -0.2425 , -0.0366 ) F(xi):  -0.6127
iteration:  10 r(i):  0.19517 xi :( -0.9371 , -0.2333 , -0.0432 ) F(xi):  -0.6151
iteration:  11 r(i):  0.18725 xi :( -0.9485 , -0.2182 , -0.0462 ) F(xi):  -0.6174
iteration:  12 r(i):  0.11001 xi :( -0.9572 ,

array([[-0.99999999],
       [-0.2       ],
       [-0.04      ]])

**Comments on Results:**

- Convergence criteria is: 100 maximum iterations or **||ri|| ≤ 10^(-8)**
- Using the initial guess as xT = [0,0,0] approxiamated value of x is **[-1.0,-0.2,0.04]**
- The converging function value is **-0.62**
- The algorithm converged in **71 iterations** using the above guess vector
- The computed values are cross-validates by checking plugging in these values given by the criteria defined above and   checking for convergence ~0.

**Comment on Method Used:**
- The conjugate gradient method works by generating a set of vectors d that are conjugate with respect to the matrix     A. That is, **dTi A dj = 0, i != j**
- The formula for αi( used as **a** here) corresponds to an **exact line search along the direction di(used as d in     the code)**